File: python/prompd/prompd_loader.py
"""
Custom Jinja2 Loader for Prompd Files

This loader enables {% include %} directives in .prmd files with compile-aware behavior:
- .prmd files: Parses and returns only the body content (frontmatter stripped)
- Other files (.md, .txt, etc.): Returns raw content as-is

Features:
- Workspace-aware path resolution (relative to source file)
- Circular include detection
- Compile-on-demand for .prmd files
"""

import threading
from pathlib import Path
from typing import Dict, Optional, Set, Tuple

from jinja2 import BaseLoader, TemplateNotFound

# Global include stack for circular dependency detection
# Keyed by compilation ID to support concurrent compilations
_global_include_stacks: Dict[str, Set[str]] = {}
_stack_lock = threading.Lock()
_compilation_counter = 0


def generate_compilation_id() -> str:
    """Generate a unique compilation ID."""
    global _compilation_counter
    with _stack_lock:
        _compilation_counter += 1
        import time

        return f"compile-{_compilation_counter}-{int(time.time() * 1000)}"


def cleanup_compilation(compilation_id: str) -> None:
    """Clean up the global stack after compilation is complete."""
    with _stack_lock:
        _global_include_stacks.pop(compilation_id, None)


class PrompdLoader(BaseLoader):
    """
    Custom Jinja2 loader that handles .prmd files specially.

    When a .prmd file is included, this loader:
    1. Parses the frontmatter
    2. Extracts only the body content
    3. Returns the body for inclusion

    This allows prompts to be composed from other prompts without
    duplicating frontmatter or metadata.
    """

    def __init__(
        self, base_dir: Path, verbose: bool = False, max_depth: int = 10, compilation_id: Optional[str] = None
    ):
        """
        Initialize the Prompd loader.

        Args:
            base_dir: Base directory for resolving relative paths
            verbose: Enable verbose logging
            max_depth: Maximum include depth to prevent infinite recursion
            compilation_id: Unique compilation ID for tracking include stack
        """
        self.base_dir = Path(base_dir)
        self.verbose = verbose
        self.max_depth = max_depth

        # Use provided compilation ID or generate a new one
        self.compilation_id = compilation_id or generate_compilation_id()

        # Initialize global stack for this compilation if needed
        with _stack_lock:
            if self.compilation_id not in _global_include_stacks:
                _global_include_stacks[self.compilation_id] = set()

    def _get_include_stack(self) -> Set[str]:
        """Get the include stack for this compilation."""
        with _stack_lock:
            return _global_include_stacks.get(self.compilation_id, set())

    def _add_to_stack(self, path: str) -> None:
        """Add a path to the include stack."""
        with _stack_lock:
            if self.compilation_id in _global_include_stacks:
                _global_include_stacks[self.compilation_id].add(path)

    def get_source(self, environment, template: str) -> Tuple[str, str, callable]:
        """
        Get the source content for a template.
        This is the main entry point called by Jinja2.

        Args:
            environment: Jinja2 environment
            template: Template name/path

        Returns:
            Tuple of (source, filename, uptodate_func)

        Raises:
            TemplateNotFound: If template cannot be found
        """
        # Resolve path relative to base directory
        resolved_path = self._resolve_path(template)
        include_stack = self._get_include_stack()

        if self.verbose:
            print(f"[include] Resolving: {template} -> {resolved_path}")

        # Check for circular includes
        if str(resolved_path) in include_stack:
            stack = " -> ".join(sorted(include_stack))
            raise TemplateNotFound(f"Circular include detected: {stack} -> {resolved_path}")

        # Check max depth
        if len(include_stack) >= self.max_depth:
            raise TemplateNotFound(
                f"Maximum include depth ({self.max_depth}) exceeded. " "Check for deep nesting or circular includes."
            )

        # Check if file exists
        if not resolved_path.exists():
            # Try with .prmd extension if not specified
            if not str(resolved_path).endswith(".prmd"):
                with_ext = resolved_path.parent / (resolved_path.name + ".prmd")
                if with_ext.exists():
                    resolved_path = with_ext
                else:
                    raise TemplateNotFound(f"Template not found: {template} (resolved to {resolved_path})")
            else:
                raise TemplateNotFound(f"Template not found: {template} (resolved to {resolved_path})")

        # Load the file
        return self._load_file(resolved_path)

    def _load_file(self, file_path: Path) -> Tuple[str, str, callable]:
        """
        Load and process a file.

        Args:
            file_path: Path to the file

        Returns:
            Tuple of (source, filename, uptodate_func)
        """
        # Add to include stack
        self._add_to_stack(str(file_path))

        # Read file content
        try:
            content = file_path.read_text(encoding="utf-8")
        except Exception as e:
            raise TemplateNotFound(f"Failed to read {file_path}: {e}") from e

        # Process based on file extension
        ext = file_path.suffix.lower()

        if ext == ".prmd":
            # Parse .prmd file and extract body only
            processed_content = self._process_prompd_file(content, file_path)
        else:
            # Return raw content for other file types
            processed_content = content

        if self.verbose:
            preview = processed_content[:50].replace("\n", "\\n")
            print(f"[include] Loaded {ext} file: {file_path} ({preview}...)")

        # Return (source, filename, uptodate_func)
        # uptodate_func returns False to always recompile (no caching)
        return (processed_content, str(file_path), lambda: False)

    def _process_prompd_file(self, content: str, file_path: Path) -> str:
        """
        Process a .prmd file: parse frontmatter and return only the body.

        Args:
            content: File content
            file_path: Path to the file (for error messages)

        Returns:
            Body content (after frontmatter)
        """
        try:
            from .parser import PrompdParser

            parser = PrompdParser()
            parsed = parser.parse_content(content)

            if self.verbose:
                param_count = len(parsed.metadata.parameters) if parsed.metadata and parsed.metadata.parameters else 0
                print(f"[include] Parsed .prmd: {file_path} ({param_count} params)")

            # Return just the body content (after frontmatter)
            return parsed.content or ""

        except Exception as e:
            # If parsing fails, log warning and return raw content minus frontmatter
            if self.verbose:
                print(f"[include] Warning: Failed to parse {file_path}: {e}")

            # Simple regex to strip frontmatter
            import re

            without_frontmatter = re.sub(r"^---\r?\n[\s\S]*?\r?\n---\r?\n?", "", content)
            return without_frontmatter

    def _resolve_path(self, name: str) -> Path:
        """
        Resolve a path relative to the base directory.

        Args:
            name: Template name/path

        Returns:
            Resolved absolute path
        """
        path = Path(name)

        # Handle absolute paths
        if path.is_absolute():
            return path

        # Handle relative paths (./file.prmd, ../file.prmd, file.prmd)
        return self.base_dir / path

    def create_child_loader(self, new_base_dir: Path) -> "PrompdLoader":
        """
        Create a child loader for nested includes.
        The child loader shares the compilation ID to use the same include stack.

        Args:
            new_base_dir: Base directory for the child loader

        Returns:
            New PrompdLoader instance
        """
        return PrompdLoader(
            base_dir=new_base_dir,
            verbose=self.verbose,
            max_depth=self.max_depth,
            compilation_id=self.compilation_id,  # Share the compilation ID
        )

    def get_compilation_id(self) -> str:
        """Get the compilation ID for this loader."""
        return self.compilation_id

File: core/src/lib/compiler/prompd-loader.ts
/**
 * Custom Nunjucks Loader for Prompd Files
 *
 * This loader enables {% include %} directives in .prmd files with compile-aware behavior:
 * - .prmd files: Parses and returns only the body content (frontmatter stripped)
 * - Other files (.md, .txt, etc.): Returns raw content as-is
 *
 * Features:
 * - Workspace-aware path resolution (relative to source file)
 * - Circular include detection
 * - Compile-on-demand for .prmd files
 */

import * as nunjucks from 'nunjucks';
import { IFileSystem } from './file-system';
import { PrompdParser } from '../parser';
import { extname, isAbsolutePosix, isPrompdFile, PROMPD_EXTENSIONS } from './path-utils';

/**
 * Global include stack shared across all loader instances.
 * This is necessary because Nunjucks creates new loader instances for nested includes,
 * but we need to track the entire include chain to detect circular dependencies.
 *
 * The stack is keyed by a unique compilation ID to support concurrent compilations.
 */
const globalIncludeStacks = new Map<string, Set<string>>();

/**
 * Generate a unique compilation ID.
 */
let compilationCounter = 0;
function generateCompilationId(): string {
  return `compile-${++compilationCounter}-${Date.now()}`;
}

/**
 * Options for the Prompd loader.
 */
export interface PrompdLoaderOptions {
  /** File system abstraction (real FS or in-memory) */
  fileSystem: IFileSystem;
  /** Base directory for resolving relative paths */
  baseDir: string;
  /** Enable verbose logging */
  verbose?: boolean;
  /** Maximum include depth to prevent infinite recursion */
  maxDepth?: number;
  /** Unique compilation ID for tracking include stack across loader instances */
  compilationId?: string;
}

/**
 * Loader result with source content and metadata.
 * Matches Nunjucks LoaderSource interface.
 */
interface LoaderResult {
  src: string;
  path: string;
  noCache: boolean;
}

/**
 * Custom Nunjucks loader that handles .prmd files specially.
 *
 * When a .prmd file is included, this loader:
 * 1. Parses the frontmatter
 * 2. Extracts only the body content
 * 3. Returns the body for inclusion
 *
 * This allows prompts to be composed from other prompts without
 * duplicating frontmatter or metadata.
 */
export class PrompdLoader extends nunjucks.Loader {
  // Required by Nunjucks loader interface - we use synchronous loading
  async = false;

  private fileSystem: IFileSystem;
  private baseDir: string;
  private verbose: boolean;
  private maxDepth: number;
  private parser: PrompdParser;
  private compilationId: string;

  constructor(options: PrompdLoaderOptions) {
    super();
    this.fileSystem = options.fileSystem;
    this.baseDir = options.baseDir;
    this.verbose = options.verbose || false;
    this.maxDepth = options.maxDepth || 10;
    this.parser = new PrompdParser();

    // Use provided compilation ID or generate a new one
    // The compilation ID links all loader instances in a single compilation
    this.compilationId = options.compilationId || generateCompilationId();

    // Initialize global stack for this compilation if needed
    if (!globalIncludeStacks.has(this.compilationId)) {
      globalIncludeStacks.set(this.compilationId, new Set());
    }
  }

  /**
   * Get the include stack for this compilation.
   */
  private getIncludeStack(): Set<string> {
    return globalIncludeStacks.get(this.compilationId) || new Set();
  }

  /**
   * Clean up the global stack after compilation is complete.
   * Should be called when the top-level compilation finishes.
   */
  static cleanupCompilation(compilationId: string): void {
    globalIncludeStacks.delete(compilationId);
  }

  /**
   * Get the source content for a template.
   * This is the main entry point called by Nunjucks.
   * Throws an error if the file is not found (required by Nunjucks interface).
   */
  getSource(name: string): LoaderResult {
    // Resolve path relative to base directory
    const resolvedPath = this.resolvePath(name);
    const includeStack = this.getIncludeStack();

    if (this.verbose) {
      console.log(`[include] Resolving: ${name} -> ${resolvedPath}`);
    }

    // Check for circular includes
    if (includeStack.has(resolvedPath)) {
      const stack = Array.from(includeStack).join(' -> ');
      throw new Error(
        `Circular include detected: ${stack} -> ${resolvedPath}`
      );
    }

    // Check max depth
    if (includeStack.size >= this.maxDepth) {
      throw new Error(
        `Maximum include depth (${this.maxDepth}) exceeded. Check for deep nesting or circular includes.`
      );
    }

    // Check if file exists
    const exists = this.fileSystem.exists(resolvedPath);
    if (typeof exists === 'object' && 'then' in exists) {
      // Async not supported in getSource - throw helpful error
      throw new Error(
        'PrompdLoader requires synchronous file system operations. Use NodeFileSystem or ensure MemoryFileSystem is pre-populated.'
      );
    }

    if (!exists) {
      // Try Prompd source extensions if none was specified (.prmd, then .md)
      if (!isPrompdFile(resolvedPath)) {
        for (const ext of PROMPD_EXTENSIONS) {
          const withExt = resolvedPath + ext;
          if (this.fileSystem.exists(withExt) === true) {
            return this.loadFile(withExt);
          }
        }
      }

      // Nunjucks expects an error to be thrown for missing files
      throw new Error(`Template not found: ${name} (resolved to ${resolvedPath})`);
    }

    return this.loadFile(resolvedPath);
  }

  /**
   * Load and process a file.
   * Note: We do NOT remove from stack after loading because Nunjucks will process
   * the returned content which may contain more includes. The stack tracks the
   * entire include chain during a single compilation.
   */
  private loadFile(filePath: string): LoaderResult {
    const includeStack = this.getIncludeStack();

    // Add to include stack - this persists for the entire compilation
    includeStack.add(filePath);

    // Read file content
    const content = this.fileSystem.readFile(filePath);
    if (typeof content === 'object' && 'then' in content) {
      throw new Error(
        'PrompdLoader requires synchronous file system operations.'
      );
    }

    // Process based on file extension
    const ext = extname(filePath).toLowerCase();
    let processedContent: string;

    if (isPrompdFile(filePath)) {
      // Parse Prompd source (.prmd or .md) and extract body only. For a .md
      // without frontmatter, processPrompdFile returns the content unchanged.
      processedContent = this.processPrompdFile(content, filePath);
    } else {
      // Return raw content for other file types
      processedContent = content;
    }

    if (this.verbose) {
      const preview = processedContent.substring(0, 50).replace(/\n/g, '\\n');
      console.log(`[include] Loaded ${ext} file: ${filePath} (${preview}...)`);
    }

    return {
      src: processedContent,
      path: filePath,
      noCache: true  // Don't cache - files may change
    };
  }

  /**
   * Process a .prmd file: parse frontmatter and return only the body.
   */
  private processPrompdFile(content: string, filePath: string): string {
    try {
      const parsed = this.parser.parseContent(content);

      if (this.verbose) {
        const paramCount = parsed.metadata?.parameters?.length || 0;
        console.log(`[include] Parsed .prmd: ${filePath} (${paramCount} params)`);
      }

      // Return just the body content (after frontmatter)
      // If no body, return empty string
      return parsed.content || '';
    } catch (error) {
      // If parsing fails, log warning and return raw content
      const errorMessage = error instanceof Error ? error.message : String(error);
      if (this.verbose) {
        console.log(`[include] Warning: Failed to parse ${filePath}: ${errorMessage}`);
      }

      // Fall back to raw content minus frontmatter
      // Simple regex to strip frontmatter
      const withoutFrontmatter = content.replace(/^---\r?\n[\s\S]*?\r?\n---\r?\n?/, '');
      return withoutFrontmatter;
    }
  }

  /**
   * Resolve a path relative to the base directory.
   */
  private resolvePath(name: string): string {
    // Handle absolute paths
    if (isAbsolutePosix(name)) {
      return name;
    }

    // Handle relative paths (./file.prmd, ../file.prmd, file.prmd)
    return this.fileSystem.join(this.baseDir, name);
  }

  /**
   * Create a child loader for nested includes.
   * The child loader shares the compilation ID to use the same include stack.
   */
  createChildLoader(newBaseDir: string): PrompdLoader {
    return new PrompdLoader({
      fileSystem: this.fileSystem,
      baseDir: newBaseDir,
      verbose: this.verbose,
      maxDepth: this.maxDepth,
      compilationId: this.compilationId  // Share the compilation ID for cycle detection
    });
  }

  /**
   * Get the compilation ID for this loader.
   * Useful for cleanup after compilation.
   */
  getCompilationId(): string {
    return this.compilationId;
  }
}

/**
 * Create a Nunjucks environment configured with the Prompd loader.
 *
 * @param options - Loader options
 * @returns Configured Nunjucks environment
 */
export function createPrompdEnvironment(options: PrompdLoaderOptions): nunjucks.Environment {
  const loader = new PrompdLoader(options);

  const env = new nunjucks.Environment(loader, {
    autoescape: false,
    throwOnUndefined: false,
    trimBlocks: true,
    lstripBlocks: true,
    tags: {
      blockStart: '{%',
      blockEnd: '%}',
      variableStart: '{{',
      variableEnd: '}}',
      commentStart: '{#',
      commentEnd: '#}'
    }
  });

  return env;
}
